### ISSUE TITLE
Stopped working as of 6-9-23

### ISSUE THREAD

**Issue by SD777x** [hwnd-git/Diablo-IV-AHK-WASD#3]
Let me first say a big thank you, this script is awesome and worked pretty decently up until last night.  Today it seems to have stopped working somewhat, I see that they did a small patch earlier so maybe something was broken. Today when I hit a WASD key it looks like the mouse is only a tiny bit ahead of the character, so to move I would have to keep repeatedly clicking on WASD.  You can see the yellow mouse circles appear very close to the character instead of how they should be a few inches away. 

Hopefully you can get it fixed, thanks much again!

**Comment by sheatsley** [comment]
Unfortunately it's not working for me. When I hold down keys my character doesn't move, but on release, I do see a click in the center.

**Comment by hwnd-git** [comment]
@SD777x 
> Edit: It's working again for some reason. Aside from this, is there any way to make this work with the arrow keys instead of WASD?

Yes it is, but it's not a matter of just few simple tweaks. Unfortunately, a substantial refactoring of the code would be needed.
Potential workaround - if you have a programmable gaming keyboard, you could try to rebind its arrow keys to 'wasd'. Or maybe use some another rebinding software to make arrows act as 'wasd'.

**Comment by hwnd-git** [comment]
@sheatsley 
> Unfortunately it's not working for me. When I hold down keys my character doesn't move, but on release, I do see a click in the center.

It will be very hard to help you, as pinpointg the exact cause is not an easy task. For starters, I would add SoundBeep command at the end of 'CallClick' function:

```
CallClick(coord)
{
	Global
	If !WinActive(appName)
		return

	xCoord := coord.x
	yCoord := coord.y
	
	ControlClick, x%xCoord% y%yCoord%, A,, L, 1, NA
	SoundBeep, 5000, 10
}
```
That way you will head a clicky sound each time the script tries to click. You will be able to tell if triggering movement by pressing one of 'wasd' actually clicks the mouse.
Then, I would check the coordinates of the click by adding a line after the soundbeep, before the bracket:
```Tooltip, `n`n`nx%xCoord%`ny%yCoord%```
It will generate small tooltip under your mouse cursor telling you the coordinates of the click (those are not apostrophes, those are the small things when you shift-press the tylda key). Press and hold one of 'wasd' buttons to check the movement clicks' coordinates (they should fluctuate a little bit, as the script sends the clicks in random time intervals as you hold the direction, but should be consistent nevertheless). As soon as you release the 'wasd' you will get the stopping click coordinate (seems like this part works for you). For consistency, you can also modify the config section:
```
xOffset := 0
yOffset := 0
(...)
sustainedMovementPeriod := 1000     ;(will trigger click every 1 second when the movement key is being held)
sustainedMovementDeviation := 0
(...)
moveCoordDeviation := 0
stopCoordDeviation := 0		
```
This should make the movement coordinates to sit exactly at the edges of the screen and match your screen resolution.
Oh, and try turning on windowed mode, not fullscreen mode in game.

Do that, and let me know the results.

**Comment by sheatsley** [comment]
@hwnd-git thanks for your help. Really appreciate it. Here's my results:
(1) _Adding SoundBeep_: I hear a sound beep _when I release W, A, S, or D_ -- this is the only time I hear a beep (I hear no beeps and my character does not move while holding down W, A, S, or D.
(2) _Adding Tooltip_: I see the tooltip appear under my cursor (wherever it may be on my screen) _when I release W, A, S, or D_ -- like the above, no tooltip appears/updates when holding down W, A, S, or D.
(3) _Modifying Config & Windowed_: No changes. I hear beeps, I see a click in the center of my screen, and a tooltip appears wherever my mouse is located on W, A, S, or D release, but nothing while any of those keys are held.

**Comment by sheatsley** [comment]
@hwnd-git thanks for your help. Really appreciate it. Here's my results:
(1) _Adding SoundBeep_: I hear a sound beep _when I release W, A, S, or D_ -- this is the only time I hear a beep (I hear no beeps and my character does not move while holding down W, A, S, or D.
(2) _Adding Tooltip_: I see the tooltip appear under my cursor (wherever it may be on my screen) _when I release W, A, S, or D_ -- like the above, no tooltip appears/updates when holding down W, A, S, or D.
(3) _Modifying Config_: Alright, this changed behavior. Now, when I press either W, A, S, or D, my character does not move, but on release, my character walks to the top left corner of my screen and the tooltip still appears wherever my cursor is, but says `x:0 y:0`.

**Comment by hwnd-git** [comment]
@sheatsley OK 1) and 2) suggest that the script does not register key-down events at all. To confirm this supposition, you could enter SoundBeep command right after:
```
w::
a:: 
s::
d::
Critical
SoundBeep
```
a) If the sound does not play at all, it means that the keys are not registered at all. That may suggest that some other software potentially interferes with AHK. Joy2Key perhaps? reWASD? Razer Synapse rebind?
b) If you can hear the beep though, then it means that the script fails somewhere during its logic. I won't be of much help for you in such case, as that would demand manual debugging to track for potential bug. I assume you are not familiar with AHK scripting language but for simple debugging, start with inserting SoundBeep here end there to see if the script runs at given line. Then you can start using tooltip or msgbox commands to check the values of given variables (the syntax is little bit convoluted for my taste, you have to wrap the variable name with %var_name%). AHK has nice documentation with examples online: https://www.autohotkey.com/docs/v1/lib/MsgBox.htm

3) This is unexpected as it broke the stopping behavior, but it should not have any influence on it (checked it on my side too). I have no idea at the moment why that might be, sorry :/

**Comment by sheatsley** [comment]
@hwnd-git I appreciate all of your help. I have this figured out now. Your script was working perfectly fine. I use [Moonlight](https://moonlight-stream.org) to stream games from my Windows machine. After directly plugging my keyboard into my machine, your script works great! So it appears that the way Moonlight sends keypresses does not allow AHK to capture pressing down. Thanks again for your help.

**Comment by SD777x** [comment]
> @SD777x
> 
> > Edit: It's working again for some reason. Aside from this, is there any way to make this work with the arrow keys instead of WASD?
> 
> Yes it is, but it's not a matter of just few simple tweaks. Unfortunately, a substantial refactoring of the code would be needed. Potential workaround - if you have a programmable gaming keyboard, you could try to rebind its arrow keys to 'wasd'. Or maybe use some another rebinding software to make arrows act as 'wasd'.

Thanks for the reply!  At this point I've just forced myself to get used to using a controller.
